paperqa/clients/openalex.py

```diff
@@ -0,0 +1,239 @@
+                "OpenAlex API returned a failed status for the query."
+            )
+
+        results_data = response_data
+        if params.get("filter") is not None:
+            results_data = results_data["results"]
+            if len(results_data) == 0:
+                raise DOINotFoundError(
+                    "OpenAlex API did not return any items for the query."
+                )
+            results_data = results_data[0]
+
+        if (
+            doi is None
+            and title
+            and strings_similarity(results_data.get("title", ""), title)
+            < title_similarity_threshold
+        ):
+            raise DOINotFoundError(
+                f"OpenAlex results did not match for title {title!r}."
+            )
+        if doi and results_data.get("doi") != doi:
+            raise DOINotFoundError(f"DOI {doi!r} not found in OpenAlex.")
+
+        return await parse_openalex_to_doc_details(results_data)
+
+
+async def parse_openalex_to_doc_details(message: dict[str, Any]) -> DocDetails:
+    """Parse OpenAlex API response to DocDetails.
+
+    Args:
+        message: The OpenAlex API response message.
+
+    Returns:
+        Parsed document details.
+    """
+    authorships = message.get("authorships") or []
+    authors = [
+        authorship.get("raw_author_name", "")
+        for authorship in authorships
+        if authorship
+    ]
+    authors = [reformat_name(author) for author in authors]
+    sanitized_authors = [
+        mutate_acute_accents(text=author, replace=True) for author in authors
+    ]
+
+    publisher = (
+        message.get("primary_location", {})
+        .get("source", {})
+        .get("host_organization_name")
+    )
+    journal = message.get("primary_location", {}).get("source", {}).get("display_name")
+
+    best_oa_location = message.get("best_oa_location")
+    pdf_url = best_oa_location.get("pdf_url") if best_oa_location else None
+    oa_license = best_oa_location.get("license") if best_oa_location else None
+
+    return DocDetails(  # type: ignore[call-arg]
+        key=None,
+        bibtex_type=BIBTEX_MAPPING.get(message.get("type", "other"), "misc"),
+        bibtex=None,
+        authors=sanitized_authors,
+        publication_date=datetime.fromisoformat(message.get("publication_date", "")),
+        year=message.get("publication_year"),
+        volume=message.get("biblio", {}).get("volume"),
+        issue=message.get("biblio", {}).get("issue"),
+        publisher=publisher,
+        issn=message.get("primary_location", {}).get("source", {}).get("issn_l"),
+        pages=message.get("biblio", {}).get("last_page"),
+        journal=journal,
+        url=message.get("doi"),
+        title=message.get("title"),
+        citation_count=message.get("cited_by_count"),
+        doi=message.get("doi"),
+        license=oa_license,
+        pdf_url=pdf_url,
+        other=message,
+    )
+
+
+class OpenAlexProvider(DOIOrTitleBasedProvider):
```

Comment:
Can you add a brief docstring on:
- What OpenAlex is
- What kinds of information it provides

Reply:
Done!